primitives/src/merkle_tree/internal.rs

```diff
@@ -544,73 +544,105 @@ where
     }
 
     /// Update the element at the given index.
-    pub(crate) fn update_internal<H, Arity>(
+    pub(crate) fn update_with_internal<H, Arity, F>(
         &mut self,
         height: usize,
         pos: impl Borrow<I>,
         traversal_path: &[usize],
-        elem: impl Borrow<E>,
-    ) -> Result<LookupResult<E, (), ()>, PrimitivesError>
+        f: F,
+    ) -> Result<(i64, LookupResult<E, (), ()>), PrimitivesError>
     where
         H: DigestAlgorithm<E, I, T>,
         Arity: Unsigned,
+        F: FnOnce(Option<&E>) -> Option<E>,
     {
         let pos = pos.borrow();
-        let elem = elem.borrow();
         match self {
             MerkleNode::Leaf {
                 elem: node_elem,
                 value,
                 pos,
-            } => {
-                let ret = ark_std::mem::replace(node_elem, elem.clone());
-                *value = H::digest_leaf(pos, elem)?;
-                Ok(LookupResult::Ok(ret, ()))
+            } => match f(Some(node_elem)) {
+                Some(elem) => {
+                    let result = (0i64, LookupResult::Ok(node_elem.clone(), ()));
+                    *value = H::digest_leaf(pos, &elem)?;
+                    *node_elem = elem;
+                    Ok(result)
+                },
+                None => {
+                    *self = MerkleNode::Empty;
+                    Ok((-1i64, LookupResult::NotFound(())))
```

Comment:
Not clear what the `LookupResult` represents either. Here it looks like the result refers to the new tree (element is `NotFound` because we just deleted it) but above we returned the old value (`node_elem`)

Reply:
Oh this is a silly bug. Will document it and fix it

Reply:
[937c573](https://github.com/EspressoSystems/jellyfish/pull/475/commits/937c57379424b074bfe5caf61d3b7538a8b8deb8)